rye/src/cli/publish.rs

```diff
@@ -68,96 +76,101 @@ pub fn execute(cmd: Args) -> Result<(), Error> {
-        write_credentials(&credentials)?;
-
-        secret
-    } else if let Some(token) = credentials
-        .get(repository)
-        .and_then(|table| table.get("token"))
-        .map(|token| token.to_string())
-        .map(escape_string)
-    {
-        let secret = Secret::new(token);
+    // Token came from file if it didn't come from the command line
+    let should_decrypt = token.is_none() && credentials.password.is_some();
 
-        maybe_decrypt(&secret, cmd.yes)?
-    } else {
-        echo!("No access token found, generate one at: https://pypi.org/manage/account/token/");
-        let token = if !cmd.yes {
-            prompt_for_token()?
+    // Token came from command line
+    let mut should_encrypt = token.is_some();
+
+    // Fallback prompts
+    if credentials.password.is_none() && !cmd.yes {
+        if credentials.username.is_none() {
+            credentials.username = prompt_username()?;
+        }
+
+        if is_unknown_repository(&repository) || is_default_repository(&repository) {
+            echo!("No access token found, generate one at: https://pypi.org/manage/account/token/");
+            credentials.password = prompt_token()?;
         } else {
-            "".to_string()
-        };
-        if token.is_empty() {
-            bail!("an access token is required")
+            credentials.password = prompt_password()?;
         }
-        let secret = Secret::new(token);
-        let maybe_encrypted = maybe_encrypt(&secret, cmd.yes)?;
-        let maybe_encoded = maybe_encode(&secret, &maybe_encrypted);
-        credentials[repository]["token"] = Item::Value(maybe_encoded.expose_secret().into());
+        should_encrypt = credentials.password.is_some();
+    }
 
-        secret
+    let passphrase = if cmd.yes || (!should_decrypt && !should_encrypt) {
+        None
+    } else if should_decrypt {
+        prompt_decrypt_passphrase()?
+    } else {
+        prompt_encrypt_passphrase()?
     };
 
-    credentials[repository]["repository-url"] = Item::Value(repository_url.to_string().into());
-    credentials[repository]["username"] = Item::Value(username.clone().into());
-    write_credentials(&credentials)?;
+    if repository.url.is_none() && !cmd.yes {
+        repository.url = prompt_repository_url()?;
+    }
+
+    // keyring fallback
+    if credentials.username.is_none() && repository.url.is_some() {
+        credentials.username = Some(DEFAULT_USERNAME.to_string())
+    }
+
+    if credentials.password.is_none() && !is_keyring_ready(&credentials, &repository) {
+        bail!(
+            "no configuration was found for repository '{}'",
+            repository.name.unwrap_or_default()
+        );
+    }
+
+    if !cmd.skip_save_credentials && repository.name.is_some() {
+        save_rye_credentials(
+            &mut credentials_file,
+            &credentials,
+            &repository,
+            should_encrypt,
+            passphrase.as_ref(),
+        )?;
+    }
```

Comment:
Note I need to test this for when repository name is not resolved.

Reply:
To address this I used `toml_edit` `Entry`s for easier management of when a write can occur using `.or_insert`.